### ISSUE TITLE
RFC 3.3/4.0: Using separated getter/setters

### ISSUE THREAD

**Issue by dereuromark** [cakephp/cakephp#7954]
Currently we use those combined setters/getters everywhere basically.
Those have three large drawbacks IMO:
They
- complicate the logic inside a method, having to separate the get vs set cases
- Make null setting impossible
- Don't go in the direction PHP goes in general, with PHP7 and stricter return types (currently always at least two: value for get and $this etc for set)

And one of the more dangerous things with the current way:
Chaining can blow up (Setters can accidently become getters resulting in fatal errors).
I have seen this myself a few times and seen it reported by people:

    $this->setterOne($value)->setterTwo(...)->...

Now if $value for some reason becomes null (sure, it shouldnt), it will fatal error as it setterOne(null)=getterOne() and therefore returning maybe a scalar value which breaks the chain hard.

We are also one of the only code bases that actually follow this combined API approach afaik.
If we want to simplify the API, extracting those into separate methods makes most sense.
It solves the above issues by design and is also more clear in its usage, using the prefixes:
```
foo() -> getFoo()
foo($x) => setFoo($x)
```
Methods are cheap and should be rather doubled as a result than possible mistakes with the internal get/set switching.

Often times we have super complicated stuff with 2nd or 3rd arguments then, because depending on the first they behave differently or not. This would be trivial with separation as those are then only applicable for the respective method.

It also allows us in the future to more easily adapt PHP7 return type hinting without having to break API everywhere. Reducing the number of possible return types should be a priority for a clean method API in 3.x to clean all that up then in 4.x as stated in the [roadmap](https://github.com/cakephp/cakephp/wiki/4.x-Roadmap).
Please note that this roadmap also already contains a note about the issues of the current API and that this is not future proof as it stands.

We can also start deprecating and using the separated setters/getters in 3.x and then remove those in 4.x.
That would ease the migration.

Most of those separations can also be automated with the upgrade app tooling I imagine.
If there is no argument it must be a setter, if there is a argument (and not null assumed) it must be a getter.
I will write a sniffer/fixer to check on any return type that is more than a single one and we can check why that is the case. 
Even `SomeObject|null` should be avoided, as null is not a valid secondary return type anymore in PHP7 strict terms, but that is a different topic alltogether (especially regarding object getters returning null fallbacks).

//EDIT: To ease migration we could also keep those deprecated combined ones in 4.x while recommending not to use them anymore. There is no need to directly throw them over board except for the maintenance overhead. But since we are not adding return typehinting any time soon (for PHP5 compatibility) there will not be any issue if we decide to keep them a while longer. 
The only important thing is that people can start to write code the *new* way asap to avoid having to refactor later. All those who want to postpone know that and can live with the consequences.
But in 5.0 we definitely should remove them then.

**Comment by ionas** [comment]
Having enjoyed some trips to the ruby world I really like combined setters/getters. It is sad that PHP did not include function parameter overloading which would have allowed us to have clean and readable code compared to the get/set verbose Java style.

But that's not where PHP seems to go for now. So I am :+1: on separating getters setters during 3.x releases, deprecating mixed methods for 4.x

I am really :+1: on making CakePHP4 rely on PHP7 features. I'd rather like to see 3.x maintained as a 5.6 release that has some backports. CakePHP2 almost died (laravel, yii, symphony) cause of its slow adoption speed.

**Comment by markstory** [comment]
I'm coming around on having separate get/set methods as well. The current combined methods make for more complex code and can get tripped up by sometimes `null` arguments. I'd also recommend that if we do go down the route of separate methods, we should also separate merge + overwrite methods.

**Comment by burzum** [comment]
I like the combined getters/setters. Can we at least keep them?

They could call the getter and setter methods internally and this way we can keep the convenient interface we have right now while still improving the code by separating the two different operations into two methods as it should be.

@ionas I don't like Yii, actually I think most people who use it are OOP newbies because the last version I've looked at looked like a pretty thin layer around standard php. I guess its popular because of that, not because of quality. Laravel and Symfony are another league but I disagree strongly with the slow adoption speed: Laravel upset it's community by making large API changes in a very short time frame. If you run a huge website or work in a large company you can't migrate every few weeks your whole app because the elite hackers of the framework thought it's just cool to be on the bleeding edge. That's usually not what larger business want, they want stability and support over a longer period of time. So if you want to satisfy the "enterprise" market don't screw with your API every few month.

**Comment by dereuromark** [comment]
Sure! keeping those around would only be a minimal overhead. They would simply forward to their new counterparts.

**Comment by ionas** [comment]
Do I understand correctly, that what missing is basically `nil` compared to `null` in PHP?

**Comment by dereuromark** [comment]
@ionas I think you are trying to take this into a really bad direction again, off-topic and totally irrelevant for what is addresses here in this scope. We are not talking about inventing PHP8 or 9 here.

**Comment by ionas** [comment]
@dereuromark Nothing bad about it. Totally in scope. AFAIK cause there is no way to differentiate between nil and null I support your concept.

**Comment by bravo-kernel** [comment]
I must admit that I am (totally) unaware of the technical implications of this proposal but as your typical end-user having a clear distinction between get() and set() methods would sure make a lot of code more understandable/intuitive so :+1: for me

**Comment by hmic** [comment]
+1+1+1 THIS is the correct solution for combined getters/setters! Null problem solved altogether!

GREAT WORK @ionas!

As CakePHP 3.2 is php-5.5 only, this should be used like *everywhere* from now on!
If support for php < 5.5 is neccessary only neccessary change would be to save the return type of func_get_args() in a temporary variable, still *very* usable!

@ionas are you going to go through the codebase yourself and setup a PR for it or leave it to others to do so (progrssively)? - I'd like to see that in the 3.2 release already!

**Comment by dereuromark** [comment]
@hmic That might solve this, but it does not solve the other issues mentioned.
So we should still split the combined ones. 
But if we can improved the code base already now without breaking BC that is sure also possible as separate task.

**Comment by markstory** [comment]
Another option is `func_num_args() === 0` or `func_num_args() === 1` for checking get/set mode.

**Comment by ionas** [comment]
After going through the core looking for `=== null`, `!== null`, `get/set`, `getter/setter`, `gets/sets`, `gets and sets` and replacing the null comparisons in https://github.com/cakephp/cakephp/pull/8051 here is my...

### Second Look
1. The speed issue is minor. It can be circumvented by using explicit getters/setters, if you care about this minor difference. @Jippy Ref: https://github.com/cakephp/cakephp/pull/8051#issuecomment-172288613
2. The overwriting issue (calling parent::) is a big issue. In future it could be circumvented by overwriting explicit getters/setters. The combined getter/setter in the child class can then stay like it is. @chinpei215 Ref: https://github.com/cakephp/cakephp/pull/8051#issuecomment-172291546
3. Sometimes it is not clear if it is just a getter/setter or also involves setting defaults if none are found, or sometimes triggering off whole processes. We should only offer combined getters/setters on easy methods, e.g. those that do precisely set protected properties or get them and do nothing beyond! A good example is `class Email` which could be improved easily offering both, combined and explicit getters/setters.

### Proposal
1. I'd love to see that we kept a non-javaish way to interact with the core framework for the most part. These parts would offer both explicit getters/setters and combined ones which just call these explicit ones. The other parts would only offer setFoo/getFoo.
2. Above is not BC and has to be moved to 4.x.
3. Right now I would like to keep the work in this PR that is BC by merging the updated doc blocks. That will help with identifying where work has to be done on the new getter/setter architecture, e.g. hopefully explicit setters/getters and combined ones for like 95% the easy cases, for 4.x if that is deemed the way to go.

Does that sound good to you, guys?

**Comment by markstory** [comment]
I'm not sure 4.x should be a big BC breaking release. 3.x was an enormous breaking change and following that up with another big breaking change soon after is not good stability and making upgrades easier. Ideally 4.x would be simply removing all the features/methods/options that were deprecated during 3.x. To that end  I would prefer if we found a non-breaking way to achieve the results we want with getters/setters.

**Comment by dereuromark** [comment]
@markstory Splitting doesn't mean the original combined one has to be removed.

@ionas The doc blocks are of course fine for now. And to take a deeper look into all non-trivial (> 2 argument) combiners.

**Comment by markstory** [comment]
@dereuromark No it does not. But that is one way someone could approach the problem. I think we have to avoid that approach and also not cause too many troubles for overridden methods.

**Comment by chinpei215** [comment]
Taking into consideration, this is a very difficult issue.
Shown below is the idea, isn't it?
```php
public function getSomething() {
	return $this->_something;
}

public function setSomething($arg) {
	$this->_something = $arg;
}

public function something($arg = null) {
	if ($arg === null) {
		return $this->setSomething($arg);
	}
	return $this->getSomething();
}

```
It seems to work even if  `something()` method is overridden. 
But it doesn't work as expected actually. 
```php
public function something($arg = null) {
	if ($arg === 'Core\Something') {
		$arg = 'App\Something';
	}
	return parent::something($arg);
}
```
In this case, the user want to use `App\Something` class instead of `Core\Something` class.
But if we call `setSomething()` in the core, the user's code would break.
```php
// $instance->something('Core\Something');
$instance->setSomething('Core\Something'); // Optimize
```
We would have to remove the combined one if we want to separate getter/setter.
So ... I am not in favor of this change at the moment.

**Comment by dereuromark** [comment]
@chinpei215 Yes, if you split, you *must* only add custom overwriting in the specific method OR keep all non-app code to the wrapper until completely removed.

So from my standpoint these are the options:

### a) Soft migration
Keep the old combiners without deprecating them really

Pros:
- Less breaking changes
- Can be automated using a script most of the way

Cons:
- Can be problematic and error prone regarding overwriting.

### b) Hard cut

Split and deprecate the combiners using trigger_error() or even Exceptions or by removing them completely.

Pros:
- Not confusing or error prone (each error message or exception tells you exactly what to use instead)
- Can be automated using a script most of the way

Cons:
- Requires migration if exception or more, otherwise tells people that keeping the deprecated code calls is possibly error prone.

### Result
I think the only option that really makes sense is b).
And for that we need the major version jump, so 4.0
It makes sense from this point, that 4.x is basically a deprecation removal, so this splitting would be the only real change happening there on top, which has a defined scope.
This makes it possible for 5.x to finally use a very clean and future proof API and everyone has time to either jump through 3.x=>4.x=>5.x faster or slower depending on their needs.

**Comment by hmic** [comment]
I do actually like the combined ones still.

 What's the problem with just keeping them in addition to providing split getters and setters?

I actually don't think that overwriting getters/setters to intercept and modify data beeing set is the right approach anyway. That's nothing than bad design! If his is neccessary to archive something because the core does it wrong in the firstplace, this is what has to be fixed instead of intercepting data IMHO.

**Comment by dereuromark** [comment]
@hmic There is no point of having them, especially with the pitfalls mentioned above.
Yes, but bad design cannot be enforced by breaking the apps.
And combined ones are a even worse concept alltogether, they neither say what they are doing nor do they reduce bugs, the opposite in fact due to the overhead logic involved.

**Comment by chinpei215** [comment]
I don't know my example code is bad design or not. Personally, I don't like inheritance. So I would use composition instead. But you can also see such a code in the core.
```php
// Cake\ORM\Association\BelongsTo.php
    public function foreignKey($key = null)
    {
        if ($key === null) {
            if ($this->_foreignKey === null) {
                $this->_foreignKey = $this->_modelKey($this->target()->alias());
            }
            return $this->_foreignKey;
        }
        return parent::foreignKey($key);
    }
```
This method wouldn't work if we separate getter/setter. But I don't think we are using bad designed framework though.

**Comment by hmic** [comment]
OK, so you want to get rid of them, why 5.0 not 4.0 and mark them deprecated in 3.3 already? Introducing the new, not BC breaking split getters/setters in 3.3 too?

One thing I do want a split setter to do is returning the set value, like the combined ones do now too :p
Any complains about this behavior?
Alternatively they should at least return true/false indicating if the value has been set

**Comment by dereuromark** [comment]
@hmic I think everyone agrees that a setter should be void or just $this (fluent interface) :)
And a getter, well no discussion needed.

**Comment by hmic** [comment]
I don't really agree.

Setting a value can go wrong, you can eighter argument throwing an exception in this case would be the correct action, or indicating that the set failed - return false!

I don't see why returning true/false would be bad, as you can just ignore it if you expect a void return :p

**Comment by dereuromark** [comment]
setting should never go wrong or should throw exceptions IMO :)
Whats the point of returning false when you dont notice because everyone treats it like void?

**Comment by hmic** [comment]
Sure, it's an argument. But as these split setters/getters would need to be implemented and are new functionality in cake, they can be implemented the way we like. Returning $this is an options as false too... Don't you think so?

**Comment by dereuromark** [comment]
Sure, but for a good API returning $this or clear exception is the way to go.

**Comment by hmic** [comment]
I'd opt for void in this case ;-)

**Comment by ionas** [comment]
> One thing I do want a split setter to do is returning the set value, like the combined ones do now too :p
Any complains about this behavior?

Thought about this too, but I do think its bad practice. While in theory setters might just set a protected property in practice they could also trigger off whole processes. If that happens what about "other" return values (like false or even status codes). If that's a no issue, then I am with you on this @hmic

**Comment by markstory** [comment]
Has anyone considered having a new opt-in classes that provide the new interface(s) we would like?  For example applications could opt-in to a new Query class that does not feature as many overloaded methods.

**Comment by dereuromark** [comment]
Since this is now part of the roadmap and somewhat already starting to get addressed even in 3.4+, I will close this now as all the facts have been gathered and we can now slowly continue the execution process.
